# **Project Plan**

## **Import Packages and Data**

In [11]:
import pandas as pd

In [12]:
try:
    contract_df = pd.read_csv('/Users/kellyshreeve/Desktop/Data-Sets/Final Project/final_provider/contract.csv', parse_dates=True)
    internet_df = pd.read_csv('/Users/kellyshreeve/Desktop/Data-Sets/Final Project/final_provider/internet.csv', parse_dates=True)
    personal_df = pd.read_csv('/Users/kellyshreeve/Desktop/Data-Sets/Final Project/final_provider/personal.csv', parse_dates=True)
    phone_df = pd.read_csv('/Users/kellyshreeve/Desktop/Data-Sets/Final Project/final_provider/phone.csv', parse_dates=True)
except:
    contract_df = pd.read_csv('/datasets/contract.csv', parse_dates=True)
    internet_df = pd.read_csv('/datasets/internet.csv', parse_dates=True)
    personal_df = pd.read_csv('/datasets/personal.csv', parse_dates=True)
    phone_df = pd.read_csv('/datasets/phone.csv', parse_dates=True)


In [13]:
print('Contract DF')
print()
print(contract_df.info())
print()

contract_df.head()

Contract DF

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   BeginDate         7043 non-null   object 
 2   EndDate           7043 non-null   object 
 3   Type              7043 non-null   object 
 4   PaperlessBilling  7043 non-null   object 
 5   PaymentMethod     7043 non-null   object 
 6   MonthlyCharges    7043 non-null   float64
 7   TotalCharges      7043 non-null   object 
dtypes: float64(1), object(7)
memory usage: 440.3+ KB
None



customerID   BeginDate              EndDate            Type  \
0  7590-VHVEG  2020-01-01                   No  Month-to-month   
1  5575-GNVDE  2017-04-01                   No        One year   
2  3668-QPYBK  2019-10-01  2019-12-01 00:00:00  Month-to-month   
3  7795-CFOCW  2016-05-01                   No        One year   
4  9237-HQITU  2019-09-01  2019-11-01 00:00:00  Month-to-month   

  PaperlessBilling              PaymentMethod  MonthlyCharges TotalCharges  
0              Yes           Electronic check           29.85        29.85  
1               No               Mailed check           56.95       1889.5  
2              Yes               Mailed check           53.85       108.15  
3               No  Bank transfer (automatic)           42.30      1840.75  
4              Yes           Electronic check           70.70       151.65

* General Info: The contract dataset contains information on customers' beginning and end date, type of plan, billing type, payment method, and charges.
* Size: This dataset contains 7043 observations and 8 columns. 
* Missing Values: There are no np.nan missing values, though further investigation is needed as there may be values recoreded as missing in another way. 
* Data Types: Total charges is read in as object, suggesting there may be some string entries in the column that will need to be removed so the data type can be properly recorded as float. Dates have not been properly parsed and should be explicitly defined as datetime.

In [14]:
print('Internet DF')
print()
print(internet_df.info())
print()

internet_df.head()

Internet DF

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5517 entries, 0 to 5516
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   customerID        5517 non-null   object
 1   InternetService   5517 non-null   object
 2   OnlineSecurity    5517 non-null   object
 3   OnlineBackup      5517 non-null   object
 4   DeviceProtection  5517 non-null   object
 5   TechSupport       5517 non-null   object
 6   StreamingTV       5517 non-null   object
 7   StreamingMovies   5517 non-null   object
dtypes: object(8)
memory usage: 344.9+ KB
None



customerID InternetService OnlineSecurity OnlineBackup DeviceProtection  \
0  7590-VHVEG             DSL             No          Yes               No   
1  5575-GNVDE             DSL            Yes           No              Yes   
2  3668-QPYBK             DSL            Yes          Yes               No   
3  7795-CFOCW             DSL            Yes           No              Yes   
4  9237-HQITU     Fiber optic             No           No               No   

  TechSupport StreamingTV StreamingMovies  
0          No          No              No  
1          No          No              No  
2          No          No              No  
3         Yes          No              No  
4          No          No              No

* General Info: The internet dataset contains information on type of internet service and whether customers have added extra protection or streaming services.
* Size: This dataset contains 5517 observations and 8 columns. 
* Missing Values: There are no np.nan missing values, though further investigation is needed as there may be values recoreded as missing in another way. 
* Data Types: All data types are read in correctly as object.

In [15]:
print('Personal DF')
print() 
print(personal_df.info())
print()

personal_df.head()

Personal DF

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   customerID     7043 non-null   object
 1   gender         7043 non-null   object
 2   SeniorCitizen  7043 non-null   int64 
 3   Partner        7043 non-null   object
 4   Dependents     7043 non-null   object
dtypes: int64(1), object(4)
memory usage: 275.2+ KB
None



customerID  gender  SeniorCitizen Partner Dependents
0  7590-VHVEG  Female              0     Yes         No
1  5575-GNVDE    Male              0      No         No
2  3668-QPYBK    Male              0      No         No
3  7795-CFOCW    Male              0      No         No
4  9237-HQITU  Female              0      No         No

* General Info: The personal dataset contains information on customers's demographics, including gender, senior status, partnership, and dependents.
* Size: This dataset contains 7043 observations and 5 columns. 
* Missing Values: There are no np.nan missing values, though further investigation is needed to determine if values have been recorded as missing in another way. 
* Data Types: All data types are correct.

In [17]:
print('Phone DF')
print()
print(phone_df.info())
print()

phone_df.head()

Phone DF

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6361 entries, 0 to 6360
Data columns (total 2 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   customerID     6361 non-null   object
 1   MultipleLines  6361 non-null   object
dtypes: object(2)
memory usage: 99.5+ KB
None



customerID MultipleLines
0  5575-GNVDE            No
1  3668-QPYBK            No
2  9237-HQITU            No
3  9305-CDSKC           Yes
4  1452-KIOVK           Yes

## **Data Distributions**

## **Project Plan**

**Prepare Data**
* Convert all column names to snake case
* Combine all datasets with an outer merge on the 'cutomer_id' column. Customers who do not have a service will be filled with np.nan for those columns.

**Create Features**
* Create a new feature named 'churned' where 0 = has not churned and 1 = has churned
* Create a new feature named 'contract_length' that contains the number of days from beginning to end of contract, for those who have churned, and the number of days from beginning of contract to latest date in the dataset, for those who have not churned.
* Create a new feature named 'begin_year' containing the year of contract start.
* Create a new feature named 'number_of_internet_services' that is the count of number of extra internet services a customer subscribes to (online_security through streaming_movies).

**Exploratory Analysis**
* Construct frequency tables and bar charts for each categorical feature.
* Construct histograms for each quantitative feature.
* Determine class balance and apply imbalanced techniques if the split is larger than 80/20.

**Build Models**
* One-Hot Encode each categorical variable
* Logistic Regression
* 


**Questions**
* If a customer is not listed on the internet or phone dataset, can it be assumed they do not have those services?